src/ast/scopes/Scope.ts

```diff
@@ -6,19 +6,19 @@ import { EntityPathTracker } from '../utils/EntityPathTracker';
 import { UNDEFINED_EXPRESSION } from '../values';
 import ArgumentsVariable from '../variables/ArgumentsVariable';
 import ExportDefaultVariable from '../variables/ExportDefaultVariable';
-import ExternalVariable from '../variables/ExternalVariable';
-import GlobalVariable from '../variables/GlobalVariable';
 import LocalVariable from '../variables/LocalVariable';
 import ThisVariable from '../variables/ThisVariable';
+import UndefinedVariable from '../variables/UndefinedVariable';
 import Variable from '../variables/Variable';
 
 export default class Scope {
 	parent: Scope | null;
 	variables: {
-		this: ThisVariable | LocalVariable;
-		default: ExportDefaultVariable;
-		arguments: ArgumentsVariable;
-		[name: string]: LocalVariable | GlobalVariable | ExternalVariable | ArgumentsVariable;
+		this?: ThisVariable | LocalVariable;
+		default?: ExportDefaultVariable;
+		arguments?: ArgumentsVariable;
+		undefined?: UndefinedVariable;
```

Comment:
I left this out previously, because `undefined` can actually be redefined as a variable identifier, so in theory we should allow it to be deshadowed.

Reply:
Ah yes, removed the type from the list. I think redeclarations should be handled properly, though, as a local declaration would of course shadow the global one now.